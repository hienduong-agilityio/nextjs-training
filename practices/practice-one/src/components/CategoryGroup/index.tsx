'use client';

// Libraries
import { usePathname, useSearchParams } from 'next/navigation';
import { twMerge } from 'tailwind-merge';

// Components
import Link from 'next/link';

// Helpers
import { capitalizeText } from '@/helpers';

// Constants
import { ALL_CATEGORIES, SEARCH_PARAMS } from '@/constants';

interface ICategoryGroupProps {
  title: string;
  items: { name: string; count: number }[];
}

export function CategoryGroup({ title, items }: ICategoryGroupProps) {
  const pathname = usePathname();
  const searchParams = useSearchParams();

  const currentCategory =
    searchParams.get(SEARCH_PARAMS.CATEGORY) ?? ALL_CATEGORIES.ALL;

  const handleCategoryClick = (itemName: string) => {
    const updatedParams = new URLSearchParams();

    // Update the category query parameter
    if (itemName !== ALL_CATEGORIES.ALL) {
      updatedParams.set(SEARCH_PARAMS.CATEGORY, itemName);
    }

    return `${pathname}?${updatedParams.toString()}`;
  };

  return (
    <div className="mb-6 bg-secondary-100 p-4 rounded">
      <h3 className="text-lg font-bold mb-4">{title}</h3>
      <ul className="space-y-2 text-sm text-gray-700">
        {items.map((item) => {
          const isActive =
            item.name === ALL_CATEGORIES.ALL
              ? currentCategory === ALL_CATEGORIES.ALL || !currentCategory
              : item.name === currentCategory;

          return (
            <li className="flex justify-between" key={item.name}>
              <Link
                href={handleCategoryClick(item.name)}
                className={twMerge(
                  'w-full block font-semibold py-2 rounded hover:bg-secondary-200',
                  isActive
                    ? 'bg-primary-500 text-primary-200'
                    : 'text-gray-800',
                )}
              >
                {capitalizeText(item.name)}
              </Link>
            </li>
          );
        })}
      </ul>
    </div>
  );
}
