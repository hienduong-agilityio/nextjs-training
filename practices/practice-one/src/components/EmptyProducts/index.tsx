import Link from 'next/link';

export const EmptyProducts = ({
  message,
  actionLabel,
  actionHref,
}: {
  message: string;
  actionLabel: string;
  actionHref: string;
}) => (
  <div className="flex flex-col items-center gap-4 py-16">
    <p className="text-xl text-gray-600">{message}</p>
    <Link
      href={actionHref}
      className="inline-block px-6 py-2 text-white bg-blue-500 rounded-lg"
    >
      {actionLabel}
    </Link>
  </div>
);
