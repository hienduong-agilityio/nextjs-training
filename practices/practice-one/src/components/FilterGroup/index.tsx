interface FilterGroupProps {
  title: string;
  items: { name: string; count: number }[];
}

export const FilterGroup = ({ title, items }: FilterGroupProps) => {
  return (
    <div className="mb-6 bg-secondary-100 p-4 rounded">
      <h3 className="text-lg font-bold mb-4">{title}</h3>
      <ul className="space-y-2 text-sm text-gray-700">
        {items.map((item) => (
          <li key={item.name} className="flex justify-between py-2">
            <button className="font-semibold hover:text-primary-200 ">
              {item.name}
            </button>
            <span>({item.count})</span>
          </li>
        ))}
      </ul>
    </div>
  );
};
