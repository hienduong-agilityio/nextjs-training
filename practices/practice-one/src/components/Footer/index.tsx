// Icon
import {
  FacebookIcon,
  LogoIcon,
  MasterCardIcon,
  PaypalIcon,
  TwitterIcon,
  VisaIcon,
  WesternIcon,
} from '@/icons';

// Mock
import { FOOTER_NAVIGATION } from '@/mocks';

// Components
import Link from 'next/link';

// Constants
import { ROUTE } from '@/constants';

export const Footer = () => {
  return (
    <footer className="w-full bg-primary-50">
      <div className="md:container px-4 pt-6 md:mx-auto sm:px-6 md:px-8 lg:px-12 xl:px-16 sm:pt-8 md:pt-12 lg:pt-16 xl:pt-20">
        <div className="py-8">
          {/* Main Footer Content */}
          <div className="flex flex-col gap-10 lg:flex-row lg:justify-between">
            {/* Logo Section */}
            <div className="max-w-[290px] flex flex-col gap-4">
              <Link href={ROUTE.PRODUCTS} className="flex items-center gap-2">
                <LogoIcon aria-label="logo" />
                <span className="text-lg font-bold sm:text-xl lg:text-2xl">
                  E-Comm
                </span>
              </Link>
              <span className="text-sm font-normal leading-normal">
                Lorem Ipsum is simply dummy text of the printing and typesetting
                industry. Lorem Ipsum has been the industry standard dummy text
                ever. Since the 1500s, when an unknown printer.
              </span>
            </div>

            <div className="flex flex-col w-full gap-8 sm:flex-row sm:justify-evenly lg:w-3/4">
              {/* Follow Us */}
              <div className="lg:max-w-[220px] flex flex-col gap-4">
                <h4 className="font-medium text-foreground-400">Follow Us</h4>
                <span className="text-sm">
                  Since the 1500s, when an unknown printer took a galley of type
                  and scrambled.
                </span>
                <div className="flex gap-4">
                  <Link href="https://www.facebook.com/" aria-label="Facebook">
                    <FacebookIcon color="#385C8E" />
                  </Link>
                  <Link href="https://www.twitter.com/" aria-label="Twitter">
                    <TwitterIcon />
                  </Link>
                </div>
              </div>

              {/* Contact Us */}
              <div className="lg:max-w-[150px] flex flex-col gap-4">
                <h4 className="font-medium text-foreground-400">Contact Us</h4>
                <address className="text-sm not-italic">
                  E-Comm, 4578 Marmora Road, Glasgow D04 89GR
                </address>
              </div>
            </div>
          </div>

          {/* Footer Links */}
          <div className="grid grid-cols-1 gap-8 pt-16 sm:grid-cols-2 lg:grid-cols-4">
            {FOOTER_NAVIGATION.map((item) => (
              <ul key={item.title}>
                <li>
                  <h4 className="mb-5 font-medium text-foreground-400">
                    {item.title}
                  </h4>
                  <ul>
                    {item.items.map((item) => (
                      <li
                        key={item.name}
                        className="mb-3 text-sm text-content2 hover:text-primary-700"
                      >
                        <Link href={item.url} className="hover:underline">
                          {item.name}
                        </Link>
                      </li>
                    ))}
                  </ul>
                </li>
              </ul>
            ))}
          </div>
        </div>

        {/* Footer Bottom */}
        <div className="flex flex-col items-center justify-between gap-6 py-6 border-t-2 border-secondary-100 lg:flex-row ">
          <p className="text-sm text-secondary-500">
            &#169; 2018 Ecommerce theme by www.bisenbaev.com
          </p>
          <div className="flex gap-2">
            <WesternIcon aria-label="Western" />
            <MasterCardIcon aria-label="MasterCard" />
            <PaypalIcon aria-label="Paypal" />
            <VisaIcon aria-label="Visa" />
          </div>
        </div>
      </div>
    </footer>
  );
};
