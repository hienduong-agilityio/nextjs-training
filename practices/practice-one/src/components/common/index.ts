export { default as Button } from './Button';
export { default as InputField } from './InputField';
export { default as Tabs } from './Tabs';
export { default as Select } from './Select';
export { default as IconButton } from './IconButton';
export * from './IconButton';
