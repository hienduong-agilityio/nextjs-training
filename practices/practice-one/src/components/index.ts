export * from './Header';
export { default as SearchBox } from './SearchBox';
export * from './SearchBox';
export * from './ProductTab';

// Common
export * from './common';
