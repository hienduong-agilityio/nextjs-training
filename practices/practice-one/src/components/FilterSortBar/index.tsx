'use client';

// Hooks
import { useSearchParams, useRouter } from 'next/navigation';
import { useEffect, useState } from 'react';

// Components
import { Select } from '@/components';

// Constants
import { SEARCH_PARAMS, SORT_PRODUCT_OPTIONS } from '@/constants';

interface FilterSortBarProps {
  itemCount: number;
  sortOptions: string[];
  showOptions: string[];
}

export const FilterSortBar = ({
  itemCount = 0,
  sortOptions = SORT_PRODUCT_OPTIONS,
  showOptions = ['6'],
}: FilterSortBarProps) => {
  const searchParams = useSearchParams();
  const router = useRouter();

  const [selectedLimit, setSelectedLimit] = useState<string>(showOptions[0]);
  const [selectedSort, setSelectedSort] = useState<string>(sortOptions[0]);

  // TODO: Remove useEffect.
  useEffect(() => {
    // Get query parameters from the URL
    const limitParam = searchParams.get(SEARCH_PARAMS.LIMIT) ?? showOptions[0];
    const sortParam = searchParams.get(SEARCH_PARAMS.SORT_BY) ?? sortOptions[0];

    setSelectedLimit(limitParam);
    setSelectedSort(sortParam);
  }, [searchParams, showOptions, sortOptions]);

  const updateQueryParams = (param: string, value: string) => {
    const currentParams = new URLSearchParams(searchParams.toString());
    currentParams.set(param, value);

    // Reset pagination if a filter changes
    if (param === SEARCH_PARAMS.LIMIT) {
      currentParams.delete(SEARCH_PARAMS.PAGE);
    }

    router.push(`?${currentParams.toString()}`);
  };

  const handleShowChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const value = event.target.value;

    setSelectedLimit(value);
    updateQueryParams(SEARCH_PARAMS.LIMIT, value);
  };

  const handleSortChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const value = event.target.value;

    setSelectedSort(value);
    updateQueryParams(SEARCH_PARAMS.SORT_BY, value);
  };

  return (
    <div className="bg-secondary-100 p-4 shadow-md md:sticky rounded top-0 z-30 w-full">
      <div className="flex flex-col sm:flex-row sm:justify-between sm:items-center space-y-4 sm:space-y-0">
        {/* Item count */}
        <div className="text-sm w-max text-gray-600">{itemCount} Items</div>

        {/* Filters */}
        <div className="flex flex-col sm:flex-row sm:items-center sm:space-x-4 space-y-4 sm:space-y-0">
          <div className="flex gap-2 items-center">
            <span>Sort</span>
            {/* TODO: Custom style for select arrow by tailwinds classes*/}
            <Select
              options={sortOptions}
              value={selectedSort}
              customClass="w-auto bg-secondary-100 capitalize"
              onChange={handleSortChange}
            />
          </div>
          <div className="flex gap-2 items-center">
            <span>Show</span>
            <Select
              options={showOptions}
              value={selectedLimit}
              customClass="w-auto bg-secondary-100"
              onChange={handleShowChange}
            />
          </div>
        </div>
      </div>
    </div>
  );
};
