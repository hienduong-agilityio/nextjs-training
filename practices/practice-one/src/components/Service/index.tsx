// Components
import { ServiceCard } from '@/components';

// Mock
import { serviceData } from '@/mocks';

export const Service = () => {
  return (
    <section className="pb-12 pt-20 lg:pb-[90px] lg:pt-[120px]">
      <div className="flex flex-wrap">
        {serviceData.map((service) => (
          <ServiceCard
            key={service.title}
            title={service.title}
            details={service.details}
            icon={service.icon}
          />
        ))}
      </div>
    </section>
  );
};
