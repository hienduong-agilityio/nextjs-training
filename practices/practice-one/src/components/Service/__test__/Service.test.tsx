// Libraries
import { render, screen } from '@testing-library/react';

// Components
import { Service } from '@/components';

// Mocks
import { serviceData } from '@/mocks';

describe('Service Component', () => {
  let container: HTMLElement | null = null;

  beforeEach(() => {
    const rendered = render(<Service />);

    container = rendered.container;
  });

  const getServiceHeadings = () => screen.getAllByRole('heading', { level: 4 });
  const getDetailTexts = () =>
    screen.getAllByText(
      'Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
    );

  it('matches the snapshot', () => {
    const { container } = render(<Service />);
    expect(container).toMatchSnapshot();
  });

  it('renders the correct number of ServiceCard components with titles', () => {
    const serviceHeadings = getServiceHeadings();
    expect(serviceHeadings).toHaveLength(serviceData.length);

    const detailTexts = getDetailTexts();
    expect(detailTexts).toHaveLength(serviceData.length);

    serviceData.forEach(({ title }) => {
      expect(screen.getByText(title)).toBeInTheDocument();
    });
  });

  it('applies correct structure and class names', () => {
    expect(container?.firstChild).toHaveClass(
      'pb-12 pt-20 lg:pb-[90px] lg:pt-[120px]',
    );

    const flexWrapper = container?.querySelector('div');
    expect(flexWrapper).toHaveClass('flex flex-wrap');
  });
});
