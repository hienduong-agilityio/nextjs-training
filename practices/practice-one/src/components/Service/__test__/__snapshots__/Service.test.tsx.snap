// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service Component matches the snapshot 1`] = `
<div>
  <section
    class="pb-12 pt-20 lg:pb-[90px] lg:pt-[120px]"
  >
    <div
      class="flex flex-wrap"
    >
      <div
        class="px-4 md:w-1/2 lg:w-1/3"
      >
        <div
          class="flex flex-col items-center p-10 bg-white rounded-md mb-9 shadow-2 hover:shadow-lg md:px-7 xl:px-10"
        >
          <div
            aria-label="Icon"
            class="flex items-center mb-11 justify-center rounded-2xl bg-primary"
          >
            <svg
              class="w-24 h-24"
              fill="none"
              viewBox="0 0 100 65"
              width="100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                transform="translate(50, 50) scale(1) translate(-50, -50)"
              >
                <path
                  d="M101.46 33.921v-.056a2.312 2.312 0 0 0-2.316-2.276h-.609l-9.41-18.18a2.312 2.312 0 0 0-2.063-1.245H71.004l.612-4.857A6.599 6.599 0 0 0 64.964.001h-49.84a2.463 2.463 0 0 0-2.438 2.22l-.503 4h28.4a3.236 3.236 0 0 1 3.252 3.34 3.446 3.446 0 0 1-3.426 3.338h-9.472v.014H3.127a2.437 2.437 0 0 0-2.422 2.358 2.288 2.288 0 0 0 2.3 2.363h36.823a3.46 3.46 0 0 1 2.94 3.532 3.693 3.693 0 0 1-3.647 3.575H13.74a2.461 2.461 0 0 0-2.448 2.389 2.31 2.31 0 0 0 2.322 2.389h25.089a3.461 3.461 0 0 1 2.987 3.527 3.692 3.692 0 0 1-3.671 3.577H6.82a2.462 2.462 0 0 0-2.447 2.389A2.31 2.31 0 0 0 6.696 41.4h2.087l-.816 8.542a6.599 6.599 0 0 0 6.651 7.305h1.802a10.404 10.404 0 0 0 10.295 8.613 11.106 11.106 0 0 0 10.746-8.613h22.467a7.612 7.612 0 0 0 4.886-1.831 6.639 6.639 0 0 0 4.651 1.831h.37A10.404 10.404 0 0 0 80.13 65.86a11.106 11.106 0 0 0 10.745-8.613h1.48a7.832 7.832 0 0 0 7.638-7.001l1.453-16.092v-.066c0-.022.006-.065.007-.099.002-.034.007-.049.007-.068ZM26.84 61.087a5.784 5.784 0 0 1-5.806-5.962 6.161 6.161 0 0 1 6.118-5.962 5.783 5.783 0 0 1 5.805 5.962 6.16 6.16 0 0 1-6.117 5.962Zm53.417 0a5.785 5.785 0 0 1-5.806-5.962 6.16 6.16 0 0 1 6.118-5.962 5.782 5.782 0 0 1 5.806 5.962 6.16 6.16 0 0 1-6.118 5.962Zm14.99-11.15a2.871 2.871 0 0 1-2.766 2.537h-1.6a10.404 10.404 0 0 0-10.183-8.08 11.109 11.109 0 0 0-10.61 8.08h-.492a2.025 2.025 0 0 1-1.524-.64 2.054 2.054 0 0 1-.503-1.592l3.008-33.316h6.905L76.39 29.058a6.6 6.6 0 0 0 6.651 7.307h13.431l-1.224 13.572Z"
                  fill="#FF6875"
                />
              </g>
            </svg>
          </div>
          <h4
            class="mb-[14px] w-max text-2xl font-semibold text-dark dark:text-white"
          >
            FREE SHIPPING
          </h4>
          <p
            class="text-center xl:w-1/2 w-full text-body-color dark:text-dark-6"
          >
            Lorem Ipsum is simply dummy text of the printing and typesetting industry.
          </p>
        </div>
      </div>
      <div
        class="px-4 md:w-1/2 lg:w-1/3"
      >
        <div
          class="flex flex-col items-center p-10 bg-white rounded-md mb-9 shadow-2 hover:shadow-lg md:px-7 xl:px-10"
        >
          <div
            aria-label="Icon"
            class="flex items-center mb-11 justify-center rounded-2xl bg-primary"
          >
            <svg
              fill="none"
              height="100"
              viewBox="0 0 80 80"
              width="100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                transform="translate(36, 36) scale(0.9) translate(-36, -36)"
              >
                <path
                  d="m55.719 17.508-3.772 4.844a28.287 28.287 0 1 1-15.782-5.888l-5.306 5.567 3.936 3.757 8.377-8.8 3.753-3.942-3.93-3.742L34.2.926l-3.755 3.942 5.728 5.456a34.398 34.398 0 0 0-6.905 68.344 34.76 34.76 0 0 0 5.42.426 34.413 34.413 0 0 0 21.03-61.586Z"
                  fill="#FF6875"
                />
                <path
                  d="M37.985 62.231V58.7c4.189-.468 8.264-3.113 8.264-8.137 0-6.825-6.347-7.607-10.982-8.178-2.963-.355-5.233-.766-5.233-2.561 0-2.497 3.52-2.766 5.03-2.766 2.245 0 4.64 1.056 5.455 2.403l.237.392 4.651-2.155-.23-.469a9.188 9.188 0 0 0-7.2-4.987v-3.112h-5.445v3.097c-5.075.754-8.09 3.565-8.09 7.596 0 6.571 5.967 7.237 10.32 7.726 3.88.458 5.686 1.416 5.686 3.018 0 3.05-4.226 3.288-5.52 3.288-2.883 0-5.658-1.438-6.468-3.349l-.2-.471-5.051 2.14.201.472c1.497 3.507 4.73 5.703 9.129 6.222v3.363h5.446Z"
                  fill="#FF6875"
                />
              </g>
            </svg>
          </div>
          <h4
            class="mb-[14px] w-max text-2xl font-semibold text-dark dark:text-white"
          >
            100% REFUND
          </h4>
          <p
            class="text-center xl:w-1/2 w-full text-body-color dark:text-dark-6"
          >
            Lorem Ipsum is simply dummy text of the printing and typesetting industry.
          </p>
        </div>
      </div>
      <div
        class="px-4 md:w-1/2 lg:w-1/3"
      >
        <div
          class="flex flex-col items-center p-10 bg-white rounded-md mb-9 shadow-2 hover:shadow-lg md:px-7 xl:px-10"
        >
          <div
            aria-label="Icon"
            class="flex items-center mb-11 justify-center rounded-2xl bg-primary"
          >
            <svg
              fill="none"
              height="100"
              viewBox="0 0 100 100"
              width="100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                transform="translate(50, 50) scale(0.9) translate(-50, -50)"
              >
                <path
                  d="M.159 71.2v18.565h67.667V71.199a7.818 7.818 0 0 0-7.807-7.807H7.966A7.818 7.818 0 0 0 .159 71.2Zm5.205 0a2.606 2.606 0 0 1 2.602-2.603h2.602v15.967H5.364V71.2Zm10.41-2.603h6.73l11.488 11.489 11.489-11.489h6.726v15.967H15.774V68.597ZM62.621 71.2v13.36h-5.204V68.593h2.602a2.606 2.606 0 0 1 2.602 2.606ZM38.12 68.597l-4.127 4.128-4.128-4.128h8.255Z"
                  fill="#FF6875"
                />
                <path
                  d="M53.464 52.978h3.956v-5.205h2.603a7.818 7.818 0 0 0 7.807-7.807V34.76a33.836 33.836 0 0 0-67.671.002v5.205a7.817 7.817 0 0 0 7.807 7.807h3.942a23.415 23.415 0 0 0 41.556 5.203Zm6.559-10.411H57.27a23.4 23.4 0 0 0 .145-2.603V34.76c0-.87-.048-1.739-.145-2.603h2.748a2.606 2.606 0 0 1 2.602 2.603v5.204a2.606 2.606 0 0 1-2.602 2.608l.005-.005ZM33.993 6.134a28.68 28.68 0 0 1 27.589 20.98 7.812 7.812 0 0 0-1.563-.158h-3.944a23.422 23.422 0 0 0-44.167 0H7.966c-.525 0-1.048.053-1.563.157a28.679 28.679 0 0 1 27.59-20.979ZM48.938 24.36a18.293 18.293 0 0 1-14.946 7.802H15.962a18.21 18.21 0 0 1 32.977-7.802ZM7.966 42.572a2.606 2.606 0 0 1-2.602-2.603v-5.204a2.606 2.606 0 0 1 2.602-2.603h2.753a23.425 23.425 0 0 0-.146 2.603v5.204c0 .87.05 1.739.146 2.603H7.966Zm7.807-2.603v-2.602h18.22a23.51 23.51 0 0 0 17.46-7.815c.505 1.691.761 3.447.76 5.211v5.205a18.111 18.111 0 0 1-1.76 7.807H41.8l-2.29-2.285a7.808 7.808 0 0 1-11.043 0l-3.68 3.68a13.013 13.013 0 0 0 17.009 1.209v2.599h4.928a18.209 18.209 0 0 1-30.95-13.01v.001Z"
                  fill="#FF6875"
                />
              </g>
            </svg>
          </div>
          <h4
            class="mb-[14px] w-max text-2xl font-semibold text-dark dark:text-white"
          >
            SUPPORT 24/7
          </h4>
          <p
            class="text-center xl:w-1/2 w-full text-body-color dark:text-dark-6"
          >
            Lorem Ipsum is simply dummy text of the printing and typesetting industry.
          </p>
        </div>
      </div>
    </div>
  </section>
</div>
`;
